### DISCUSSION TITLE
Has anyone tried Dolly-like models?

### DISCUSSION THREAD

**Discussion by kha84** [ggml-org/llama.cpp#569] [Ideas]
I just watched the latest video of my favorite youtuber - https://www.youtube.com/watch?v=AWAo4iyNWGc&t=14s and was wondering, if someone has already quantized & converted one of these to be compatible with llama.cpp?
The beauty of Dolly-like models is that they're based on open source [gpt-j-6B from EleutherAI](https://huggingface.co/EleutherAI/gpt-j-6B), so noone will be hunting us for using them without an ask.

**Comment by pikalover6** [comment]
[ggml](https://github.com/ggerganov/ggml) already supports gpt-j, you should just be able to convert and quantize them.

**Comment by gorborukov** [comment]
I did not succeed with the [convert-h5-to-ggml.py](https://github.com/ggerganov/ggml/blob/master/examples/gpt-j/convert-h5-to-ggml.py) script and the model [Dolly_GPT-J-6b](https://huggingface.co/TehVenom/Dolly_GPT-J-6b). Script thinks for a while and crashes with the status "Killed".
![Screenshot from 2023-03-28 01-02-58](https://user-images.githubusercontent.com/2666067/228077306-f907123c-abbc-4c5e-95c7-3e7200bdfe38.png)

**Comment by kha84** [comment]
> I did not succeed with the [convert-h5-to-ggml.py](https://github.com/ggerganov/ggml/blob/master/examples/gpt-j/convert-h5-to-ggml.py) script and the model [Dolly_GPT-J-6b](https://huggingface.co/TehVenom/Dolly_GPT-J-6b). Script thinks for a while and crashes with the status "Killed". ![Screenshot from 2023-03-28 01-02-58](https://user-images.githubusercontent.com/2666067/228077306-f907123c-abbc-4c5e-95c7-3e7200bdfe38.png)

Prob out-of-memory killer played his part? Check journalctl / syslog and try to monitor system resources while running it

**Reply by gorborukov** [reply]
Yes, there was a problem with the OOM. I created a 16gb swap and was able to convert the model. I was even able to generate text with this model using gpt-j inference from ggml repo. It remains to figure out how to run it under the llama.cpp (i am getting "Bad magic" error) and preferably quantize it.
![Screenshot from 2023-03-28 14-12-11](https://user-images.githubusercontent.com/2666067/228219346-13736f4a-d95b-4e4f-b2c3-3b209598b316.png)

**Reply by iamfaith** [reply]
Do you quantize into 4-bit model？

**Reply by gorborukov** [reply]
No, I tried the model and came to the conclusion that it is significantly worse than the llama in many ways. This is where my experiments ended, I think quantization would make things even worse.

**Reply by simonw** [reply]
Dolly 2 looks like it could be a LOT more exciting: it's based on pythia-12B and uses a fine tuning set that isn't license encumbered:  https://huggingface.co/databricks/dolly-v2-12b and https://www.databricks.com/blog/2023/04/12/dolly-first-open-commercially-viable-instruction-tuned-llm

**Reply by drdaffey** [reply]
> Dolly 2 looks like it could be a LOT more exciting: it's based on pythia-12B and uses a fine tuning set that isn't license encumbered: https://huggingface.co/databricks/dolly-v2-12b and https://www.databricks.com/blog/2023/04/12/dolly-first-open-commercially-viable-instruction-tuned-llm

If someone is going to be converting and quantizing this, I would love to try it on llama.cpp. Exciting!

**Reply by ggerganov** [reply]
This model is GPT-NeoX architecture. Is it much different from GPT-J? I think they should be pretty similar.

Pinging @ravenscroftj and @saharNooby as they have recently used `ggml` to successfully implement inference of a new model - in case they are interested in adding support for Dolly v2. Of course, everyone is welcome to give it a try. I'm currently trying to focus on the SAM implementation, so it will take me some time to come back to this

**Reply by ghost** [reply]
Considering how this model was released today it might be worth getting a sense of its capabilities before putting in the dev time. Pythia can be run without quantization on CPU, it's just slow (I run the 2.8B version on 16G RAM). There are open models like the Cerebras one which generated a lot of hype but ultimately performed poorly to the point where no one uses it.

**Reply by simonw** [reply]
I'm pretty impressed by it so far - a few examples:

```pycon
>>> print(prompt("Who is Neil Armstrong?"))
```
> Neil Armstrong was an American astronaut, fighter pilot, and test pilot, and a retired United States Navy officer. He was the first person to step onto the Moon's surface, and was followed by Buzz Aldrin on the Apollo 11 mission in 1969. Armstrong and Aldrin collected scientific samples and conducted experiments while on the Moon's surface for about two hours, before lifting off to rejoin the Apollo Lunar Module Pilot Michael Collins in orbit, where they remained for seven days while preparing for another lunar landing attempt and capturing images of the Earth to send to people on Earth. This was considered a major milestone in human history and was a significant precursor to the United States' domination of space for the next 20 years.
```pycon
>>> print(prompt("Markdown list of albums by Cher"))
Here is a list of albums by Cher

- Hole In the Pages
- Love Siegal
- The Son of No One
- It's a Man's World
- Love Always
- The Magic of Motivation
- Fantasy
- Life and how to live it
- Bittersweet White Light
- Love Calls
- The Dance
- Love’s Swelled Eyes
- Dancing with Ghosts
- Believe
- Clarity
- Cher at the Royal Albert Hall
- Cher at the Sydney Opera House
- Dancing in the Street
- The sauve face
- Chastain
```
Here's how I'm running it at the moment: https://til.simonwillison.net/llms/dolly-2

**Reply by leroycep** [reply]
I was able to convert it running using https://github.com/NolanoOrg/cformers

I don't think I figured out how to hook up the AutoTokenizer to the `interface.py` script in `cformers`, but I was able to manually convert the tokens using the databricks AutoTokenizer,

```python
>>> # convert text to tokens
>>> from transformers import AutoTokenizer
>>> tokenizer = AutoTokenizer.from_pretrained("databricks/dolly-v2-12b", padding_size="left")
>>> tokenizer.encode("### Instruction: Write a short poem about wavelengths of light ### Response:")
[50278, 19566, 247, 2159, 19361, 670, 25738, 273, 1708, 209, 50279]
```

```bash
~/git/3_resources/ai/cformers/cformers/cpp/main gptneox -m ~/git/3_resources/ai/ggml/build/models/dolly-v2-12b/ggml-model-q4_1.bin --prompt "50278 19566 247 2159 19361 670 25738 273 1708 209 50279" --seed 1337 --threads 6 --n_predict 100
```

```python
>>> # convert tokens to text
>>> for t in map(int, "50278  19566  247  2159  19361  670  25738  273  1708  209  50279  13  309  858  15  187  253  2316  369  3644  285  1335  2490  672  642  11260  21758  4543  10654  43256  2490  3345  352  369  2360  673  535  50277  27  50276  42  12293  562  715  253  2806  1255  50254  50270  35  609  3174  275  253  5412  3216  50259  3220  2798  7171  9086  13  346  2513  436  2471  865  50264  34  6627  8325  432  6706  274  50263  2809  531  949  253  43580  50258  35338  479  1928  4999  285  11874  50265  1909  604  271  23087  3261  387  479  342  10540  50268  1898  1160  619  2360  673  3150  535  50277  27  50276  3  1147  310".split()):
...    print(tokenizer.decode(t))
...
>>>

```

Torrent Links:
- magnet:?xt=urn:btih:19d66ec7b5af922e849c1a3b2702de6c21a4dd9c&dn=dolly-v2-12b-f16.ggml
- magnet:?xt=urn:btih:cb1cfa34fec50b4806efbc2e633f3d814b8aab3f&dn=dolly-v2-12b-q4_0.ggml
- magnet:?xt=urn:btih:bd67894b2df80eeae1ff5087b072cb76ddd15166&dn=dolly-v2-12b-q4_1.ggml

**Reply by maddes8cht** [reply]
Currently only the f16 download has working peers to download from...

**Reply by morpheus2448** [reply]
I've added the those torrents and will seed for a few days to give others a chance.

Thanks @leroycep!

**Reply by leroycep** [reply]
No problem! Though at the rate the files are uploading at it might be faster for people to run the conversion themselves :sweat_smile: I had to give my computer a lot of swap (initially tried 30 GiB, but that wasn't enough. Went a bit overkill and increased it to 100 GiB), but then [convert_gptneox_to_ggml.py](https://github.com/NolanoOrg/cformers/blob/master/cformers/cpp/converters/convert_gptneox_to_ggml.py) worked. Once I had done that I used [quantize_gptneox](https://github.com/NolanoOrg/cformers/blob/master/cformers/cpp/quantize_gptneox.cpp)

**Reply by maddes8cht** [reply]
At the current rate my bittorrent client estimates it will take 5 days to complete....
Maybe it speeds up when there are more parts available

**Reply by maddes8cht** [reply]
Why are there 
`dolly-v2-12b-q4_0.ggml` and 
`dolly-v2-12b-q4_1.ggml` ?
Shouldn't the `ggjt` format files be in one .bin file?
And shouldn't a 12B size quantized fit into one 8 GB file?

**Reply by leroycep** [reply]
Oh, those are different quantizations. I don't know what the difference is between q4_0 and q4_1 is, but you should only need one of them. `f16` is pre-quantization so I wouldn't recommend downloading it unless you want to quantize it yourself.

This is my first seeding a torrent, so I may be doing something wrong. I'll see if there's anything I can do to upload faster

**Reply by JarbasAl** [reply]
also seeding here, but perhaps upload somewhere we can direct download? I often use https://anonfiles.com/ when i need to host this sort of stuff

**Reply by ghost** [reply]
Since Dolly 2 is an openly licensed model you should be able to put it on Hugging Face without issues as well.

**Reply by ggerganov** [reply]
Just to highlight that there soon will be significant improvements in the 4-bit quantization quality - both `q4_0` and `q4_1`.
~So keep in mind that whatever `ggml` models you generated, you will likely want to regenerate them when the updates are merged.~

You can follow the progress #951 

Edit: actually, #951 will not lead to data format changes. It will just improve the quality of the eval with the existing data. So model re-generation will not be necessary. Still, there might be some follow-up PRs where we will change the data format to gain even more 4-bit accuracy

**Reply by leroycep** [reply]
Good to know.

I'm trying out uploading to huggingface, but it looks like it will face the same issue of slow upload speeds. I may try running the quantization on a server with a better connection if I have the time.

**Reply by morpheus2448** [reply]
Both 4-bit quantized models are seeding, the fp16 version is at 66%. Haven't had a chance to run them yet.

@maddes8cht 
q4_1 is the 'better' 4-bit quantization
q4_0 is smaller/faster/less accurate

**Reply by maddes8cht** [reply]
q4_0 and q4_1 model finished download, try to start each of them with llama.cpp main.exe on Windows with the line
```
main.exe" -m ./models/Dolly-v2/dolly-v2-12b-q4_0.ggml.bin  -n 256  --repeat_penalty 1.0 --color --instruct --threads 8 --ctx_size  2048
```
getting an error:

```
main: seed = 1681480528
llama.cpp: loading model from ./models/Dolly-v2/dolly-v2-12b-q4_0.ggml.bin
error loading model: missing tok_embeddings.weight
llama_init_from_file: failed to load model
main: error: failed to load model './models/Dolly-v2/dolly-v2-12b-q4_0.ggml.bin'
```

**Reply by leroycep** [reply]
I have uploaded the `q4_0` model to huggingface: https://huggingface.co/geemili/dolly-v2-12b/tree/main

Unfortunately, it appears that the current quantization algorithm makes the model significantly worse.

Changes made to the cformers repo to run the model:
```patch
diff --git a/cformers/chat.py b/cformers/chat.py
index b9df603..b433ffa 100644
--- a/cformers/chat.py
+++ b/cformers/chat.py
@@ -1,6 +1,6 @@
 import argparse
 from interface import AutoInference as AI
-model_map = {'pythia': 'OpenAssistant/oasst-sft-1-pythia-12b', 'bloom': 'bigscience/bloom-7b1', 'gptj': 'EleutherAI/gpt-j-6B'}
+model_map = {'pythia': 'OpenAssistant/oasst-sft-1-pythia-12b', 'bloom': 'bigscience/bloom-7b1', 'gptj': 'EleutherAI/gpt-j-6B', 'dolly2': 'databricks/dolly-v2-12b'}
 
 parser = argparse.ArgumentParser()
 parser.add_argument("-p", "--prompt", nargs="+",help="Enter a value for the prompt")
@@ -8,6 +8,16 @@ parser.add_argument("-t", "--tokens",help="Number of tokens to generate",type=in
 parser.add_argument("-m", "--model", help="Specify a Model", choices=model_map.keys(),default="pythia")
 args = parser.parse_args()
 
+
+DOLLY2_PROMPT = """Below is an instruction that describes a task. Write a response that appropriately completes the request.
+
+### Instruction:
+{}
+
+### Response:
+"""
+
+
 def generate(prompt,arg=args):
     print('Model is '+arg.model)
     if arg.model == 'pythia':
@@ -16,6 +26,8 @@ def generate(prompt,arg=args):
         x = ai.generate(""+prompt+"", num_tokens_to_generate=arg.tokens)
     elif arg.model == 'gptj':
         x = ai.generate(""+prompt+"", num_tokens_to_generate=arg.tokens)
+    elif arg.model == 'dolly2':
+        x = ai.generate(DOLLY2_PROMPT.format(prompt), num_tokens_to_generate=arg.tokens)
     else:
         x = ai.generate("<|prompter|>"+prompt+"<|endoftext|><|assistant|>", num_tokens_to_generate=arg.tokens)
     return x
diff --git a/cformers/interface.py b/cformers/interface.py
index 8821ee2..97fbe59 100644
--- a/cformers/interface.py
+++ b/cformers/interface.py
@@ -136,6 +136,9 @@ MAP_MODEL_TO_URL = { # Replace "/" with "-.-" in the model name
     'OpenAssistant/oasst-sft-1-pythia-12b': ModelUrlMap(
         cpp_model_name="gptneox",
         int4_fixed_zero="https://huggingface.co/ayushk4/OpenAssistant-.-oasst-sft-1-pythia-12b/resolve/main/int4_fixed_zero.bin"),
+    'databricks/dolly-v2-12b': ModelUrlMap(
+        cpp_model_name="gptneox",
+        int4_fixed_zero="file:///home/geemili/git/3_resources/ai/ggml/build/models/dolly-v2-12b/ggml-model-f16.bin"),
 }
 
 class AutoInference:
@@ -163,7 +166,7 @@ class AutoInference:
             wget.download(self.model_url, self.model_save_path, bar=bar_progress)
 
             print("Download complete!")
-            compare_file_hash_sha256(self.model_save_path, self.model_url.replace("resolve", "blob"))
+            # compare_file_hash_sha256(self.model_save_path, self.model_url.replace("resolve", "blob"))
 
```

`q4_0` output:
```
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Write a python script to calculate fibbonacci

### Response:
from functools import lru_cache
import itertool as IT
print(lazy = LRU(n=3))  # n is the number of workers in your cluster 
futures = { }   ## this part caches results for fast execution later on 

def fib(val):

    if val == 0 or 1: return
        
    print("Calculating Fibonacci sequence value:", end=" ")
    timeitResults()
```

`q4_1` output:
```
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Write a python script to calculate fibbonacci

### Response:
def fact(x): 
    if x == 0: return 1  # this if statment will execute only once
            return float("{0} * {1}+_".format(x-1,x-2)) # This line executes the desired result. See format string for another way to define a new number

 fact(5)                                            #executing the function 

Output: 50

### End not defined after using if statment else block

### End!
```

`f16` output (this took forever to run on my computer):
```
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Write a python script to calculate fibbonacci

### Response:
def fibonaccis(n): 
    a, b = 0, 1  # Starting values for recurrence relation
    for i in range(2 <= n < 1000]:   
        a,b=b,a+b   # Calculate next Fib number using standard formula
        yield a     # Return calculated value

### End.fibonaccis(15) # Print first 15 fib numbers returned by the function 

Output 
[0 1 1 2 3 5 8
```

**Reply by gorborukov** [reply]
> q4_0 and q4_1 model finished download, try to start each of them with llama.cpp main.exe on Windows with the line
> 
> ```
> main.exe" -m ./models/Dolly-v2/dolly-v2-12b-q4_0.ggml.bin  -n 256  --repeat_penalty 1.0 --color --instruct --threads 8 --ctx_size  2048
> ```
> 
> getting an error:
> 
> ```
> main: seed = 1681480528
> llama.cpp: loading model from ./models/Dolly-v2/dolly-v2-12b-q4_0.ggml.bin
> error loading model: missing tok_embeddings.weight
> llama_init_from_file: failed to load model
> main: error: failed to load model './models/Dolly-v2/dolly-v2-12b-q4_0.ggml.bin'
> ```

Same with newer 7b and 3b models. Seems to need some modifications to the conversion script for use dolly-v2 with llama.cpp.

**Reply by leroycep** [reply]
@gorborukov You can use the cformers gptneox conversion script ( https://github.com/NolanoOrg/cformers/blob/master/cformers/cpp/converters/convert_gptneox_to_ggml.py ) to convert the models

**Reply by gorborukov** [reply]
> @gorborukov You can use the cformers gptneox conversion script ( https://github.com/NolanoOrg/cformers/blob/master/cformers/cpp/converters/convert_gptneox_to_ggml.py ) to convert the models

I used this script, it is ok for cformers inference, but llama.cpp gives an error `error loading model: missing tok_embeddings.weight`

**Reply by maddes8cht** [reply]
So at the moment Dolly V2 does not work with Llama.cpp yet

**Reply by Ghatage** [reply]
@leroycep I'm wondering if I should try to use float32 during quantization of `databricks/dolly-v2-12b`. Perhaps that would give better accuracy (but require 2x the memory)

**Reply by Ghatage** [reply]
Just checking back, I converted the new dolly model with using float32 for the weights. The accuracy has greatly improved compared to the float16 linked above. However the model size grown 6 times the original (7gb -> 47 gb).

**Reply by leroycep** [reply]
It looks like the dolly (and pythia/GPTNeoX) were trained using 16 bit, so if the accuracy is improving, it's likely that only the computations need to changed, and not the stored values.

- 12b config shows `torch_dtype` as `bfloat16`: https://huggingface.co/databricks/dolly-v2-12b/blob/main/config.json
- GPTNeoX was trained using `fp16`: https://huggingface.co/docs/transformers/model_doc/gpt_neox
- The pythia dataset uses fp16 precision: https://github.com/EleutherAI/pythia/blob/main/models/12B/pythia-12b.yml#L61-L69

I've started work on separating the GPTNeoX runner from the cformers library into a [standalone executable]( https://github.com/leroycep/dolly). I'm using the Zig language, and I've tweaked the library to `mmap` the file, which makes the file load much faster. I haven't gotten evaluation working yet. There have been changes to GGML that break the cformers code for GPTNeoX, and I don't understand enough about GGML or GPTNeoX to fix them. I think I'll work through [Karpathy's Neural Network course](https://karpathy.ai/zero-to-hero.html) before trying to continue this effort.

**Reply by ghost** [reply]
I think there'll be a lot more interest on these courses going forwards with all these new models coming out. The other applications didn't really do it for me but with LLMs I feel that I've been bitten by the AI bug...

**Reply by ggerganov** [reply]
GPT-NeoX basic inference is now demonstrated in the `ggml` repo, via the new StableLM Alpha models:

https://github.com/ggerganov/ggml/tree/master/examples/stablelm

**Comment by jon-chuang** [comment]
I will investigate this as I would like to see foundation models and their variants outside of llama be blessed with the gift of ggml.
